### ISSUE TITLE
parallelize histogram

### ISSUE THREAD

**Issue by psteinb** [sqeazy/sqeazy#26]

**Comment by xSooDx** [comment]
Hi @psteinb 

I would like to help with the parallelize issues. I have some experience with parallel programming with openMP.

**Comment by psteinb** [comment]
Hi @xSooDx, thanks for the offer. I am curious that someone stumbled upon this package. 
Note that this library is not production ready (as you may have noticed). 

I am currently in the process of parallelising [hist_impl.hpp](https://github.com/sqeazy/sqeazy/blob/parallelize/src/cpp/src/hist_impl.hpp) and [hist utils](https://github.com/sqeazy/sqeazy/blob/parallelize/src/cpp/src/encoders/histogram_utils.hpp). This library is bound to openmp 2 as it supports gcc, clang and MSVC. the latter has not picked up any recent openmp standard AFAIK. With regards to the histogram implementation, I am currently checking if one can come up with an implementation that is SIMD-threads hybrid. My current implementation running on a 2-core ivy bridge laptop gives the following:

```
------------------------------------------------------------------------------------
Benchmark                                             Time           CPU Iterations
------------------------------------------------------------------------------------
BM_serial_fill<std::uint8_t>/33554432          57675737 ns   57211075 ns         12   559.332MB/s
BM_serial_fill<std::uint16_t>/33554432         57291959 ns   56881313 ns         12   562.575MB/s
BM_serial_SIMD_fill<std::uint8_t>/33554432     16393989 ns   16240595 ns         42   1.92419GB/s
BM_serial_SIMD_fill<std::uint16_t>/33554432    39498317 ns   39120043 ns         18   817.995MB/s
BM_omp_parallel_fill<std::uint8_t>/33554432    16166473 ns   15762661 ns         45   1.98253GB/s
BM_omp_parallel_fill<std::uint16_t>/33554432   22705994 ns   21110725 ns         33   1.48029GB/s
```
this computed the histogram from a `33554432` elements large vector of uniformly diced `uint8_t` or `uint16_t` elements. If you can come up with an implementation that can beat that, feel free to send a PR.

**Comment by psteinb** [comment]
implemented parallel hist fill with 6eea387a56628557e76d0025c9cdaa964dd0cf9f, propagated into quantiser with 22eb49347e70d8052c77c76dabf706ac90950fa1

**Comment by psteinb** [comment]
finished with 3c624bb
